gpt_json/streaming.py

```diff
@@ -0,0 +1,78 @@
+import json
+from dataclasses import dataclass
+from enum import Enum
+from typing import Any, Generic, Optional, Type, TypeVar
+
+from pydantic import BaseModel, create_model
+
+from gpt_json.transformations import JsonFixEnum, fix_truncated_json
+
+SchemaType = TypeVar("SchemaType", bound=BaseModel)
+
+class StreamEventEnum(Enum):
+    OBJECT_CREATED = "OBJECT_CREATED"
+    KEY_UPDATED = "KEY_UPDATED"
+    KEY_COMPLETED = "KEY_COMPLETED"
+    AWAITING_FIRST_KEY = "AWAITING_FIRST_KEY"
+
+
+@dataclass
+class StreamingObject(Generic[SchemaType]):
+  updated_key: str | None
+  value_change: str | None
+  event: StreamEventEnum
+  partial_obj: SchemaType
+
+  schema_model: Type[SchemaType] = None
+     
+  def __class_getitem__(cls, item):
+    new_cls = super().__class_getitem__(item)
+    new_cls.schema_model = item
+    return new_cls
+
+
+def prepare_streaming_object(schema_model: SchemaType, curr_partial_raw: dict[str, Any], prev_partial: 'StreamingObject[SchemaType]', proposed_event: StreamEventEnum) -> StreamingObject[SchemaType]:
+   # compute which key was most recently updated 
+   raw_recent_key = list(curr_partial_raw.keys())[-1] if curr_partial_raw else None
+   updated_key = raw_recent_key if raw_recent_key in schema_model.__fields__ else None
```

Comment:
When would the raw recent not be in the model schema? If GPT is hallucinating results?

Reply:
or, if the key has been partially streamed. e.g.:

{
"student_model" : "blah"
}

`student` is a separate token, and so at some point we see `curr_partial_raw` like this:

{
"student": ""
}